### PR TITLE
Some ocamltest cleanups

### PR DESCRIPTION
**Pull request by nojb**
Some cleanups extracted from #9596. It can be reviewed commit by commit. The commits are all self-explanatory except maybe "Use sum type...".

The idea is not to use objects and to put all the logic concerning ocaml tools (compilers, toplevels and ocamldoc) in the same place. Apart from the simpler language constructs used, it makes it easier to follow (and to find) if one has all the logic in the same place. And since the set of tools is pretty much fixed, we don't really have a use for the extensibility provided by objects.

All in all, -72 lines of code.

### CODE REVIEW

File: ocamltest/main.ml
Comment (gasche) [inline ocamltest/main.ml:208]: I tried to reconstruct the reasoning that this is safe. I believe it goes as follows: some parts of ocamltest do run `chdir` (the `cd` action and the `setup_*` actions), but in this piece of code the only place where they can run is guarded within a `with_chdir` wrappers that correctly resets the current directory.

File: ocamltest/filecompare.ml
Comment (gasche) [inline ocamltest/filecompare.ml]: I think this would be more readable (to me!) with the

```
Sys.with_input_file ~bin:true file1 @@ fun ic1 ->
Sys.with_input_file ~bin:true file2 @@ fun ic2 ->
``` 

idiom. (Recently I've been using a `( let@ )` binding operator for this, but this would be too daring.)

(There are several other occurrences of your double-indentation-bump approach below in the diff; so if you were to agree, it would be nice to have a look at them too.)

**Review by gasche — APPROVED** [review]
The commits generally look nice to me (I already did a pre-review in #9596), thanks! I made a nitpicky style comment.

File: ocamltest/main.ml
Comment (nojb) [reply ocamltest/main.ml:208]: That's exactly it. In fact it was me that introduced the `with_chdir` and I remember that I realised at the time that this `chdir` could go, but somehow I didn't get around to doing it.

File: ocamltest/filecompare.ml
Comment (nojb) [reply ocamltest/filecompare.ml]: Nice! Amended as suggested. (And also added one more commit applying the same style to the `Sys.with_chdir` discussed above.)

**Comment by gasche** [comment]
My comments have been addressed. (I think this would deserve a Changes entry, in particular the object removal is a non-trivial refactoring.)

On the other hand, there seems to be a testsuite failure, which I suppose was introduced by a regression in this code. I would look at a regression in the `reference_file` computation for toplevel tests.

File: ocamltest/ocaml_toplevels.ml
Comment (gasche) [inline ocamltest/ocaml_toplevels.ml:46]: This is slightly different from the `ocaml_compilers.ml` override, in that here the backend is used instead of the target.

Comment (gasche) [reply ocamltest/ocaml_toplevels.ml:46]: In fact I believe that the difference comes from the way `tool#reference_file env prefix` (the `default` choice) uses the `directory` variable, which for the toplevel is equal to `"ocaml"`.

Comment (nojb) [reply ocamltest/ocaml_toplevels.ml:46]: Indeed! Fix pushed (+ Changes entries).

**Comment by nojb** [comment]
Thanks a lot for the review; I pushed one more commit sharing the logic that handles exit codes when calling external processes.

**Comment by nojb** [comment]
> I find the code duplication unfortunate here. Before you try having a more parametrized version of "the mother of all reference-files computations", are we sure that the non-default case is actually used in the wild? It would be easy to check this, as our wild is not so wild: just remove the `else` branch and run the testsuite locally. If it passes, we can get rid of that logic (which looks to me like it is not so appropriate for toplevel tests anyway).

Ok, just launched the testsuite, will report back.

**Comment by nojb** [comment]
> Ok, just launched the testsuite, will report back.

OK, it looks to be used in a several places, eg `typing-private/private.ml`: https://github.com/ocaml/ocaml/tree/b2f467abc0d2caaf40b6446a567f968ce935247f/testsuite/tests/typing-private

But I think we can just "fix" this by changing ".compilers" to ".ocaml" in the reference files. I will give it a shot.

**Comment by nojb** [comment]
Yes, only a handful of tests were affected. I renamed the corresponding reference files and the whole testsuite passed. So I went ahead and did the change. (I also fixed a small mistake in the previous commit about the exit status logic).

**Comment by nojb** [comment]
Currently one can specify a number of bytes/lines to skip when comparing files. This quantity is attached to the comparison tool used, which allows impossible states to be represented (eg specifying a number of lines to skip when comparing binary files). I pushed a commit that instead attaches this quantity to the filetype being compared, so that only a number of bytes (resp. lines) can be specified when comparing binary (resp. text) files. I also moved the logic that disables the comparison of binaries under Windows/flambda to a more fitting place.

**Comment by gasche** [comment]
I suspect that you recreating a new monster PR :-)

I reviewed all commits up to "Simplify logic for toplevel reference files", to the exception of "Share run_cmd exit status logic" which I haven't had time to look at yet. This is the part that I approved, and I think it is basically ready for merging (modulo some git history cleanup).

I haven't looked at the other commits yet (including the run_cmd exit status one), and I don't know when/if I will do that part. They look like unrelated refactoring that could go in a separate PR (possibly on top of the first one, if there are cleanup-dependencies.)

**Comment by nojb** [comment]
> I reviewed all commits up to "Simplify logic for toplevel reference files", to the exception of "Share run_cmd exit status logic" which I haven't had time to look at yet. This is the part that I approved, and I think it is basically ready for merging (modulo some git history cleanup).

> I haven't looked at the other commits yet (including the run_cmd exit status one), and I don't know when/if I will do that part. They look like unrelated refactoring that could go in a separate PR (possibly on top of the first one, if there are cleanup-dependencies.)

OK. I will take that commit and the later ones and put them in a separate PR. I had included them here because they are very simple refactoring commits, as unrelated to anything else as the existing ones in this PR.

**Comment by nojb** [comment]
OK I force-pushed and removed all commits after "toplevel reference file logic" (+ the `run_cmd` one) and will put them in a new PR. Since the remaining commits have been approved, am planning to merge after CI passes. Thanks a lot for the quick review!

**Comment by nojb** [comment]
(I will clean up the history before merging)

**Comment by Armael** [comment]
(I haven't followed the details of the patch, but for a PR that has been submitted on a Sunday, I think it might be nicer to at least wait until Monday before merging, so that other interested people can have a chance of looking at it?)

**Comment by nojb** [comment]
> (I haven't followed the details of the patch, but for a PR that has been submitted on a Sunday, I think it might be nicer to at least wait until Monday before merging, so that other interested people can have a chance of looking at it?)

Sure, sounds fine! Thanks.

**Comment by xavierleroy** [comment]
It would be seriously rude, bordering on misconduct, to merge any of this without @shindere 's approval.

**Comment by shindere** [comment]
Sorry @xleroy, I realise it is my bad for not being prompt enough.

I will start tomorrow by examining the ocamltest-related PRs.

Sorry all for my recent lack of efficiency. Will do my best to catch-up
as promptly as possible in order not to hinder any development efforts.

**Comment by gasche** [comment]
Well of course there is no need to apologize for not being available to review something that was submitted less than two days ago, especially on the week-end, because there is absolutely nothing wrong with that.

**Comment by nojb** [comment]
> Well of course there is no need to apologize for not being available to review something that was submitted less than two days ago, especially on the week-end, because there is absolutely nothing wrong with that.

I wrote to Sébastien directly to clear up any potential confusion on this matter, but I would also like to second @gasche's comment. Cheers!

**Comment by shindere** [comment]
What I understand from this PR is that it has two main things: making
ocamlsrcdir global and removing the object-oriented logic.

I am unsure about the first change because it feels to me it
implies that the source directory will be computed at compile-time
rather than at execution time. In particular, the environment variable
loses of its interest with this change because it won't be possible
to override the predefined setting at execution time. I am not sure
this is a desirable behaviour.

Regarding the removal of the object-oriented logic, well it all depends
what is going to happen with ocamltest. At the time when I designed this
tool, my hope was that it could be used to test tools other than the
compiler itself, through a plugins mehcanism. As a consequence, I tried to
define structures that could then be re-used and extended by plugins, hence
the introduction of concepts like tools and compilers.

One of my plans was to go even further with the object-oriented logic
by introducing a state object. That state would be defined by the core
and would contain at least the environment and the log stream.

Then plugins could extend that state. The ocaml (built-in) plugin, for
instance, would add the ocamlsrcdir variable or function to the state.

It feels to me this PR does not really follow this line of thought.

I'm happy to discuss the architecture of ocamltest, though, and still hope
it can become a wider used tool.

**Comment by nojb** [comment]
> I am unsure about the first change because it feels to me it
> implies that the source directory will be computed at compile-time
> rather than at execution time. In particular, the environment variable
> loses of its interest with this change because it won't be possible
> to override the predefined setting at execution time. I am not sure
> this is a desirable behaviour.

No, the source directory will be computed, *once*, when the program starts (and the environment variable will be taken into account). What is no longer possible is to change its value during the execution of tests. I can't imagine any situation where that could be useful... can you think of any?

> Regarding the removal of the object-oriented logic, well it all depends
> what is going to happen with ocamltest. At the time when I designed this
> tool, my hope was that it could be used to test tools other than the
> compiler itself, through a plugins mehcanism. As a consequence, I tried to
> define structures that could then be re-used and extended by plugins, hence
> the introduction of concepts like tools and compilers.

Yes, I can see that. But note that what I changed is all internal to the "ocaml" plugin, so in principle my changes do not compromise the possibility of an extensible architecture.

> One of my plans was to go even further with the object-oriented logic
> by introducing a state object. That state would be defined by the core
> and would contain at least the environment and the log stream.

I don't think this is a good idea: using objects complicates matters and makes it harder to figure out the flow of the code. There are situations where they can be useful (typically involving open recursion and overriding specific behaviour in subclasses), but I don't feel this is one of them. Also, the less state in the code, the better.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2020/05/25 10:03 +0000):
> > I am unsure about the first change because it feels to me it
> > implies that the source directory will be computed at compile-time
> > rather than at execution time. In particular, the environment variable
> > loses of its interest with this change because it won't be possible
> > to override the predefined setting at execution time. I am not sure
> > this is a desirable behaviour.
>
> No, the source directory will be computed, *once*, when the program
> starts (and the environment variable will be taken into account). What
> is no longer possible is to change its value during the execution of
> tests. I can't imagine any situation where that could be useful... can
> you think of any?

Not really. I may have designed things that way because I had this wrong
belief so it's indeed certainly better to do it as you suggest. Thanks a
lot.

> > Regarding the removal of the object-oriented logic, well it all depends
> > what is going to happen with ocamltest. At the time when I designed this
> > tool, my hope was that it could be used to test tools other than the
> > compiler itself, through a plugins mehcanism. As a consequence, I tried to
> > define structures that could then be re-used and extended by plugins, hence
> > the introduction of concepts like tools and compilers.
>
> Yes, I can see that. But note that what I changed is all internal to
> the "ocaml" plugin, so in principle my changes do not compromise the
> possibility of an extensible architecture.

I think I left the concepts in the plugin because there was no other
use-case at that time but my idea was to move them to the core as soon
as there would be another use-case so that they could be shared.

> > One of my plans was to go even further with the object-oriented logic
> > by introducing a state object. That state would be defined by the core
> > and would contain at least the environment and the log stream.
>
> I don't think this is a good idea: using objects complicates matters
> and makes it harder to figure out the flow of the code. There are
> situations where they can be useful (typically involving open
> recursion and overriding specific behaviour in subclasses), but I
> don't feel this is one of them. Also, the less state in the code, the
> better.

Well by "state" I just ment a place where all the useful information
would be gathered to be passed to the different functions. Typically,
many functions take the log and env parameters. I wanted to replace them
by a "state" parameter that would have contianed both the log and the
environemnt, and that could be extended by plugins to ocntain whatever
informatiin would be useful to them, in addition to log and env. Can you
think of a nice way to achieve this without using objects and classes?

**Comment by nojb** [comment]
> > Regarding the removal of the object-oriented logic, well it all depends > what is going to happen with ocamltest. At the time when I designed this > tool, my hope was that it could be used to test tools other than the > compiler itself, through a plugins mehcanism. As a consequence, I tried to > define structures that could then be re-used and extended by plugins, hence > the introduction of concepts like tools and compilers. Yes, I can see that. But note that what I changed is all internal to the "ocaml" plugin, so in principle my changes do not compromise the possibility of an extensible architecture.
>
> I think I left the concepts in the plugin because there was no other use-case at that time but my idea was to move them to the core as soon as there would be another use-case so that they could be shared.

OK. If a more general type needs to be introduced later, the existing sum type can be rethought, or wrapped in the more general type, etc. I don't think it is a good idea to design code for a hypothetical situation which is not yet known.

> > One of my plans was to go even further with the object-oriented logic > by introducing a state object. That state would be defined by the core > and would contain at least the environment and the log stream. I don't think this is a good idea: using objects complicates matters and makes it harder to figure out the flow of the code. There are situations where they can be useful (typically involving open recursion and overriding specific behaviour in subclasses), but I don't feel this is one of them. Also, the less state in the code, the better.
>
> Well by "state" I just ment a place where all the useful information would be gathered to be passed to the different functions. Typically, many functions take the log and env parameters. I wanted to replace them by a "state" parameter that would have contianed both the log and the environemnt, and that could be extended by plugins to ocntain whatever informatiin would be useful to them, in addition to log and env. Can you think of a nice way to achieve this without using objects and classes?

Again it is hard to discuss in hypotheticals... As a general rule, whatever you want to do with objects can be done (often in a simpler way) by just using functions. The way to do it in a particular situation would depend on the specific kind of information and the use made of it.

Having said that I don't think putting `env` and `log` together is the direction we want to go in. In fact, my hope is to go in the *opposite* direction: do the environment and variable processing first (this part does not require access to `log`), and generate a description of the side effects to perform, but delaying their execution until later (at which point the `log` would become relevant). 

This kind of split is nice because the environment/variable processing functions (what is today called `Actions.t`) become pure functions, which means easier to reason about. And having a stand-alone description of the side-effects opens the door to new possibilities: different kinds of logging/tracing, analysis of the dependencies and/or outputs of the side effects, etc.

**Comment by shindere** [comment]
Have you also considered that scripts for instance can also modify
environments?

Doesn't this break the separation you'd like to introduce?

**Comment by nojb** [comment]
> Have you also considered that scripts for instance can also modify environments?
>
> Doesn't this break the separation you'd like to introduce?

Yes, I saw this. Unless I am mistaken this is only used in the `run_script` and `run_hook` actions. However, it doesn't look like this particular feature is used at all in the `testsuite` today. I see two possibilities:

- Getting rid of this functionality: do you know of a good use-case for it?
- Incorporating this possibility into the design: for example, the execution of a side effect could return a set of environment modifiers in addition to a result, which would then be reinjected in the environment before running the following action, etc.

In the second case, we would only be able to deal with actions one at a time, since the result of one action may affect the following one. I don't think this negates the benefits of the environment-handling/side-effect split.

**Comment by shindere** [comment]
I think youor second case corresponds to what is currently done, no?

The fact that the feature is used or not is, in my opinion, not the most
important point here. The most important point was that external actions
should be as expressive as those written in OCaml.

**Comment by nojb** [comment]
> I think youor second case corresponds to what is currently done, no? The fact that the feature is used or not is, in my opinion, not the most important point here.

Yes, but this issue is orthogonal to having a split between the processing of environments/variables and performing side effects.

>  The most important point was that external actions should be as expressive as those written in OCaml.

Personally I'm not sure this is a good idea (if we wanted this we could use OCaml as scripting language), but in any case this is a different discussion.

**Comment by nojb** [comment]
After some private discussion, I realise the rationale for these changes was not articulated or argued as clearly as it could have been. Unfortunately I am extremely short on time at the moment, and I will not be able to do more on that front. On that note, I am closing this PR. Thanks everyone for the discussion. Cheers!

**Comment by gasche** [comment]
Come on. Whether we like object-oriented programming or not, many of the changes in PR are *obviously good ideas*:
- use with_{input,output}_file: https://github.com/ocaml/ocaml/pull/9597/commits/caea1476acef754a400d4ca54d55c0052a358c62
- use Filename.quote_command: https://github.com/ocaml/ocaml/pull/9597/commits/d08b9babca67801c1e49e4325612ad691b8b3bad
- Use `Sys.{win32,...}` instead of `Sys.os_type`: https://github.com/ocaml/ocaml/pull/9597/commits/0b90828a3f1ca74229ce6087527671c7850067c7

We have a problem if the only person motivated to work on the `ocamltest` codebase except @shindere is discouraged from making changes, to the point that those changes proposals are lost. Sharing expertise on the codebase is very important in the long term.

Personally I think that this argument should be convincing enough to do whatever with the objects. But if we do decide that it is very important to keep objects in the codebase for potential future (unknown) extensibility, maybe we could get rid of this particular change and go through the rest of the commits?

**Comment by shindere** [comment]
I also liked the \ocamlsrcdir`change.

It may be right that pre-designing ofr an unknown extension is bad, but
since it's actually there I find it too bad to just get rid of it.

So yes, I think the PR without the removal of object-oriented aspects
may be nice to have.

**Comment by gasche** [comment]
@nojb would you be willing to re-send this PR without the object-oriented change? If not, would it work for you if I did the rebase and resubmission?

**Comment by nojb** [comment]
> @nojb would you be willing to re-send this PR without the object-oriented change? If not, would it work for you if I did the rebase and resubmission?

To be honest I am a tad busy at the moment, but no problem if you want to pick up the relay.

**Comment by gasche** [comment]
Done in #9604.

**Pull request closed without merging**
